### PR TITLE
Concurrent ERC2771

### PR DESCRIPTION
**Pull request by bennoprice**
# Relay Concurrent ERC2771

## Motivation

In addition to the existing nonce-based replay protection for ERC2771 Relay methods we would like to allow concurrent execution using hash-based replay protection.

## Solution

To support this we add two new Relay smart contracts (and corresponding abstract/interface contracts):

- `GelatoRelayConcurrentERC2771.sol`
- `GelatoRelay1BalanceConcurrentERC2771.sol`

### Roles and permissions

- [ ] It's Permissionless!
- [x] Roles and permissions listed below:

Same as other relay contracts: "onlyGelato"

### External Contracts

- [x] No external contracts.
- [ ] External contracts listed below:

### Deployment

Is this a new deployment or an upgrade to a deployed upgradable contract?

- [x] New deployment.
- [ ] Upgrade to a deployed upgradable contract:

### FUNDS AND PRIVILEGES

Does the contract hold any funds or privileges?

- [x] No funds, allowance, or privileges.
- [ ] Some:

## Checklist

- [x] I have performed a self-review of my code
- [x] If it is a new feature, I have added thorough tests.
- [ ] I have run your code through basic code analysis tools (Slither)
- [x] Solidity does not emit ANY compiler warnings.
- [x] The code is well-documented
- [x] The code has been deployed and tested on a forking network

### CODE REVIEW

File: contracts/GelatoRelay1BalanceConcurrentERC2771.sol [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+        _requireUserDeadline(
+            _call.userDeadline,
+            "GelatoRelay1BalanceConcurrentERC2771.sponsoredCallConcurrentERC2771:"
+        );
+
+        bytes32 callHash = _hashSponsoredCallConcurrentERC2771(_call);
+
+        // For the user, we enforce hash-based replay protection
+        _requireHash(
```
Comment (gitpusha) [inline contracts/GelatoRelay1BalanceConcurrentERC2771.sol:76]: rename to `_requireReplayProtection`

File: contracts/GelatoRelayConcurrentERC2771.sol [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+        _requireUserDeadline(
+            _call.userDeadline,
+            "GelatoRelayConcurrentERC2771.callWithSyncFeeConcurrentERC2771:"
+        );
+
+        bytes32 callHash = _hashCallWithSyncFeeConcurrentERC2771(_call);
+
+        // For the user, we enforce hash-based replay protection
+        _requireHash(
```
Comment (gitpusha) [inline contracts/GelatoRelayConcurrentERC2771.sol:73]: rename

File: deploy/GelatoRelay1BalanceConcurrentERC2771.deploy.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+import { sleep } from "../src/utils";
+import { getAddresses } from "../src/addresses";
+import { setCode } from "@nomicfoundation/hardhat-network-helpers";
+
+const func: DeployFunction = async (hre: HardhatRuntimeEnvironment) => {
+  const { deploy } = deployments;
+  const {
+    deployer: hardhatAccount,
+    relayERC2771Deployer,
+    devRelayERC2771Deployer,
```
Comment (gitpusha) [inline deploy/GelatoRelay1BalanceConcurrentERC2771.deploy.ts:13]: We have to remove those later and replace with new deployer keys

**Review by jack-the-pug — APPROVED** [review]
LGTM!

File: deploy/GelatoRelay1BalanceConcurrentERC2771.deploy.ts (hunk 3 above)
Comment (denis-pingin) [reply deploy/GelatoRelay1BalanceConcurrentERC2771.deploy.ts:13]: The code should be updated to use a new deployment key for both new contracts and add `dependencies` config

File: deploy/GelatoRelay1BalanceConcurrentERC2771.deploy.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
+import { deployments, getNamedAccounts } from "hardhat";
+import { HardhatRuntimeEnvironment } from "hardhat/types";
+import { DeployFunction } from "hardhat-deploy/types";
+import { sleep } from "../src/utils";
+import { getAddresses } from "../src/addresses";
+//import { setCode } from "@nomicfoundation/hardhat-network-helpers";
```
Comment (denis-pingin) [inline deploy/GelatoRelay1BalanceConcurrentERC2771.deploy.ts:6]: Commented out code should be either removed or added back, depending on what makes sense

**Review by denis-pingin — CHANGES_REQUESTED** [review]

**Review by denis-pingin — APPROVED** [review]

**Pull request merged**